airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -111,19 +111,34 @@ def next(self) -> Optional[Mapping[str, Any]]:
         """
 
         try:
-            if self._current_slice is None:
-                self._current_slice = self._get_next_slice()
-                state_for_slice = self._cursor.select_state(self._current_slice)
-                if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
-                    return None
-                else:
-                    return self._current_slice
             if self._read_state_from_cursor:
-                state_for_slice = self._cursor.select_state(self._current_slice)
-                if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
-                    self._current_slice = self._get_next_slice()
+                if self._current_slice is None:
+                    next_slice = self._get_next_slice()
+                    state_for_slice = self._cursor.select_state(self._current_slice)
+                    if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
+                        next_candidate_slice = None
+                        has_more = True
+                        while has_more:
+                            next_candidate_slice = self._get_next_slice()
+                            state_for_slice = self._cursor.select_state(next_candidate_slice)
+                            has_more = state_for_slice == {"__ab_full_refresh_sync_complete": True}
+                        self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=next_candidate_slice.partition)
+                    else:
+                        self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=next_slice.partition)
                 else:
-                    self._current_slice = StreamSlice(cursor_slice=state_for_slice or {}, partition=self._current_slice.partition)
+                    state_for_slice = self._cursor.select_state(self._current_slice)
+                    if state_for_slice == {"__ab_full_refresh_sync_complete": True}:
+                        # Skip every slice that already has the terminal complete value indicating that a previous attempt
+                        # successfully synced the slice
+                        next_candidate_slice = None
+                        has_more = True
+                        while has_more:
```

Comment:
This is preferable to do within the external method, if possible, since we don't return anything, it's perfectly aligned.

Reply:
i'm not sure extracting this actually makes it more readable. I experimented a bit with this:
```
    def _find_next_unfinished_slice(self) -> StreamSlice:
        next_candidate_slice = None
        has_more = True
        while has_more:
            next_candidate_slice = self._get_next_slice()
            state_for_slice = self._cursor.select_state(next_candidate_slice)
            has_more = state_for_slice == FULL_REFRESH_COMPLETE_STATE
        return next_candidate_slice
```

It does mostly work, but we do actually need to pass back both `state_for_slice` in addition to `next_candidate_slice`. We could return it as a tuple, but it just adds more layers to where the variables need to be passed around and used. For a relatively small block of code, i think leaving it as is is cleaner

Reply:
No problem, but if it's crashed at certain point it's easier to find the actual root cause, from the method name rather than get familiar with the pretty complex logic of the whole method. Not blocking on this!